overlord/fdestate/dbx.go

```diff
@@ -0,0 +1,568 @@
+	if err != nil {
+		t.Errorf("cannot complete post update reseal: %v", err)
+	}
+
+	if err != nil {
+		op.SetFailed(
+			fmt.Sprintf("cannot complete post update reseal: %v, while completing due to external request", err))
+	} else {
+		op.SetStatus(DoneStatus)
+	}
+
+	if updateErr := updateExternalOperation(st, op); updateErr != nil {
+		t.Logf("cannot update operation status: %v", updateErr)
+		return updateErr
+	}
+
+	if err == nil {
+		// update task status before unlocking
+		t.SetStatus(state.DoneStatus)
+	}
+
+	return err
+}
+
+func (m *FDEManager) doEFISecurebootDBUpdatePrepareCleanup(t *state.Task, tomb *tomb.Tomb) error {
+	st := t.State()
+
+	st.Lock()
+	defer st.Unlock()
+
+	chgID := t.Change().ID()
+	return withFdeState(st, func(fde *FdeState) (modified bool, err error) {
+		for idx, op := range fde.PendingExternalOperations {
+			logger.Debugf("cleaning up external operation for change %v", op.ChangeID)
+			if op.ChangeID == chgID {
+				fde.PendingExternalOperations = append(fde.PendingExternalOperations[:idx],
+					fde.PendingExternalOperations[idx+1:]...)
+
+				val := st.Cached(dbxUpdatePrepareSyncKey{})
+				if val != nil {
+					syncChs := val.(map[string]chan struct{})
+					delete(syncChs, chgID)
+				}
+
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+}
+
+func isEFISecurebootDBUpdateBlocked(t *state.Task) bool {
+	extChg, err := findFirstExternalOperationByChangeID(t.State(), t.Change().ID())
+	if err != nil {
+		// error obtaining data from the state, which does not mean the
+		// operation isn't there, best case, leave it running and wait for the
+		// change to abort
+		return true
+	}
+
+	if extChg == nil {
+		// no operation, then why were we called?
+		return true
+	}
+
+	switch extChg.Status {
+	case CompletingStatus, AbortingStatus:
+		// operation states that unblock the efi-secureboot-update-db task
+		return false
+	default:
+		return true
+	}
+}
+
+func dbxUpdateAffectedSnaps(t *state.Task) ([]string, error) {
+	// TODO check if we have sealed keys at all
+
+	// DBX updates cause a reseal, so any snaps which are either directly
+	// measured or their content is measured during the boot will count as
+	// affected
+
+	// XXX this effectively blocks updates of gadget, kernel & base until the
+	// change completes
+
+	return fdeRelevantSnaps(t.State())
+}
+
+func checkDBXChangeConflicts(st *state.State) error {
+	// TODO check if we have sealed keys at all
+
+	snaps, err := fdeRelevantSnaps(st)
+	if err != nil {
+		return err
+	}
+
+	if len(snaps) == 0 {
+		return nil
+	}
+
+	// make sure that are no other DBX changes in progress
+	op, err := findFirstPendingExternalOperationByKind(st, "fde-efi-secureboot-db-update")
+	if err != nil {
+		return err
+	}
+
+	if op != nil {
+		return &snapstate.ChangeConflictError{
+			ChangeKind: "fde-efi-secureboot-db-update",
+			Message:    "cannot start a new DBX update when conflicting actions are in progress",
```

Comment:
I suppose that if fwupd for some reason fail to do a cleanup, but does not restart, we get into that situation. And I suppose it is better to fail here. The only way for a user to get unblocked here is to restart fwupd. Am I correct?